src/store_client.cc

```diff
@@ -491,34 +512,40 @@ store_client::fileRead()
               copyInto.data,
               copyInto.length,
               copyInto.offset + mem->swap_hdr_sz,
-              mem->swap_hdr_sz == 0 ? storeClientReadHeader
-              : storeClientReadBody,
+              expectHeader() ? storeClientReadHeader : storeClientReadBody,
               this);
 }
 
 void
 store_client::readBody(const char *, ssize_t len)
 {
-    int parsed_header = 0;
-
     // Don't assert disk_io_pending here.. may be called by read_header
     flags.disk_io_pending = false;
     assert(_callback.pending());
-    debugs(90, 3, "storeClientReadBody: len " << len << "");
+    debugs(90, 3, "storeClientReadBody: len " << len << " offset " << copyInto.offset);
 
     if (len < 0)
         return fail();
 
-    const auto rep = entry->mem_obj ? &entry->mem().baseReply() : nullptr;
-    if (copyInto.offset == 0 && len > 0 && rep && rep->sline.status() == Http::scNone) {
-        /* Our structure ! */
-        if (!entry->mem_obj->adjustableBaseReply().parseCharBuf(copyInto.data, headersEnd(copyInto.data, len))) {
-            debugs(90, DBG_CRITICAL, "Could not parse headers from on disk object");
-        } else {
-            parsed_header = 1;
+    auto headerParsed = !expectHeader();
+    if (!headerParsed) {
+        initReplyBuffer();
+        replyBuffer->append(copyInto.data, len);
+        replyBuffer->terminate();
+        auto error = Http::scNone;
+        auto &adjustableReply = entry->mem_obj->adjustableBaseReply();
+        if (adjustableReply.parse(replyBuffer->buf, replyBuffer->size, 0, &error)) {
+            headerParsed = true;
+            assert(adjustableReply.pstate == Http::Message::psParsed);
```

Comment:
I would also add an assert that relates to the new state management of the store client itself:

```suggestion
            assert(adjustableReply.pstate == Http::Message::psParsed);
            assert(!expectingHttpHeader()); // paranoid
```